File: rust/cudf-sys/cpp/binaryop.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION.
// SPDX-License-Identifier: Apache-2.0

#include "cudf-sys/src/binaryop.rs.h"
#include "cudf-sys/src/lib.rs.h"
#include "cudf-sys/helpers.hpp"

#include <cudf/binaryop.hpp>

namespace cudf_sys {

std::unique_ptr<Column> binary_operation_columns(
    cudf::column_view const& lhs,
    cudf::column_view const& rhs,
    BinaryOperator op,
    int32_t output_type_id,
    std::size_t stream) {
  return COL(cudf::binary_operation(lhs, rhs,
      ENUM<cudf::binary_operator>(static_cast<int32_t>(op)), DT(output_type_id), S(stream)));
}

std::unique_ptr<Column> binary_operation_column_scalar(
    cudf::column_view const& lhs,
    Scalar const& rhs,
    BinaryOperator op,
    int32_t output_type_id,
    std::size_t stream) {
  return COL(cudf::binary_operation(lhs, rhs.inner(),
      ENUM<cudf::binary_operator>(static_cast<int32_t>(op)), DT(output_type_id), S(stream)));
}

std::unique_ptr<Column> binary_operation_scalar_column(
    Scalar const& lhs,
    cudf::column_view const& rhs,
    BinaryOperator op,
    int32_t output_type_id,
    std::size_t stream) {
  return COL(cudf::binary_operation(lhs.inner(), rhs,
      ENUM<cudf::binary_operator>(static_cast<int32_t>(op)), DT(output_type_id), S(stream)));
}

int32_t binary_operation_fixed_point_scale(int32_t op, int32_t left_scale, int32_t right_scale) {
  return cudf::binary_operation_fixed_point_scale(
      ENUM<cudf::binary_operator>(op), left_scale, right_scale);
}

bool is_supported_binaryop(int32_t out_type_id, int32_t lhs_type_id, int32_t rhs_type_id, int32_t op) {
  return cudf::binops::is_supported_operation(
      DT(out_type_id), DT(lhs_type_id), DT(rhs_type_id), ENUM<cudf::binary_operator>(op));
}

}  // namespace cudf_sys

File: rust/cudf-sys/src/binaryop.rs
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION.
// SPDX-License-Identifier: Apache-2.0

#[cxx::bridge(namespace = "cudf_sys")]
pub mod ffi {
    unsafe extern "C++" {
        include!("cudf-sys/lib.hpp");

        // Shared types from the core bridge.
        type Column = crate::ffi::Column;
        type Scalar = crate::ffi::Scalar;
        type BinaryOperator = crate::ffi::BinaryOperator;
        #[namespace = "cudf"]
        type column_view = crate::ffi::column_view;

        // -- Binary operations --

        /// Binary operation between two column_views.
        fn binary_operation_columns(
            lhs: &column_view,
            rhs: &column_view,
            op: BinaryOperator,
            output_type_id: i32,
            stream: usize,
        ) -> Result<UniquePtr<Column>>;

        /// Binary operation between a column_view and a Scalar.
        fn binary_operation_column_scalar(
            lhs: &column_view,
            rhs: &Scalar,
            op: BinaryOperator,
            output_type_id: i32,
            stream: usize,
        ) -> Result<UniquePtr<Column>>;

        /// Binary operation between a Scalar and a column_view.
        fn binary_operation_scalar_column(
            lhs: &Scalar,
            rhs: &column_view,
            op: BinaryOperator,
            output_type_id: i32,
            stream: usize,
        ) -> Result<UniquePtr<Column>>;

        // -- Binary: fixed_point_scale / is_supported_operation --

        /// Compute the output scale for a fixed-point binary operation.
        fn binary_operation_fixed_point_scale(op: i32, left_scale: i32, right_scale: i32) -> i32;

        /// Check if a binary operation is supported for the given types.
        fn is_supported_binaryop(
            out_type_id: i32,
            lhs_type_id: i32,
            rhs_type_id: i32,
            op: i32,
        ) -> bool;
    }
}
